docs/src/main/tut/chisel3/chisel3-vs-compat.md

```diff
@@ -0,0 +1,203 @@
+safer and more modern Chisel 3 semantics.
+
+### Connections
+
+#### Compatibility mode
+
+In compatibility mode, both connection operators `:=` and `<>` are actually the same.
+They both are bidirectional and noncommutative--the right side will be treated as
+the source and the left side will be treated as the sink.
+
+```scala
+import Chisel._
+class MyModule extends Module {
+  val io = new Bundle {
+    val in = Decoupled(UInt(width = 4)).flip
+    val out = Decoupled(UInt(width = 4))
+  }
+  io.out <> io.in // This works
+  io.out := io.in // This is equivalent
+  io.in <> io.out // This will error
+}
+```
+For Bundles, elements with matching names will be connected and unmatching elements
+will be ignored. For example:
+```scala
+import Chisel._
+class BundleA extends Bundle {
+  val foo = UInt(width = 8)
+}
+class BundleB extends Bundle {
+  val foo = UInt(width = 8)
+  val bar = UInt(width = 8)
+}
+class MyModule extends Module {
+  val io = new Bundle {
+    val in = (new BundleA).asInput
+    val out = (new BundleB).asOutput
+  }
+  io.out <> io.in
+  // Equivalent to
+  io.out.foo := io.in.foo
+  // bar is ignored because it doesn't match
+} 
+```
+
+Furthermore, error detection and reporting is defered to FIRRTL compilation.
+
+#### Chisel 3
+
+In Chisel 3, `:=` is refered to as "monoconnect" and `<>` is called "biconnect".
+
+* Monoconnect
+`:=` treats everything on the left-hand side as a sink, even if the type is
+bidirectional. This means it cannot be used to drive bidirectional output ports,
+but can be used to drive wires from bidirectional inputs or outputs to "monitor"
+the the full aggregate, ignoring directions.
+
+* Biconnect
+`<>` performs bidirectional connections and is commutative. At least one of the
+arguments must be a port.
+It will determine the correct leaf-level connections based on the directions of
+its port arguments.
+It cannot be used to connect two components _inside_ of a module.
+
+```scala
+import chisel3._
+import chisel3.util._
+class MyModule extends Module {
+  val io = IO(new Bundle {
+    val in = Flipped(Decoupled(UInt(4.W)))
+    val out = Decoupled(UInt(4.W))
+    val x = Output(UInt(8.W))
+  })
+  io.out <> io.in // This works
+  io.in <> io.out // So does this
+  io.out := io.in // Error, cannot drive io.out.ready
+
+  val w = Wire(Decoupled(UInt(4.W)))
+  w := io.in  // This works, w can be seen as "monitoring" io.in
+  4.U <> io.x // This also works but is stylistically suspect
+}
+```
+In contrast to compatibility mode, every field of two connected Bundles must match.
+For example:
+```scala
+import chisel3._
+class BundleA extends Bundle {
+  val foo = UInt(8.W)
+}
+class BundleB extends Bundle {
+  val foo = UInt(8.W)
+  val bar = UInt(8.W)
+}
+class MyModule extends Module {
+  val io = new Bundle {
+    val in = Input(new BundleA)
+    val out = Output(new BundleB)
+  }
+  io.out <> io.in // This is an error because io.in.bar doesn't match
```

Comment:
So how to do this properly assuming the user wants the same code to be generated?

Reply:
`Chisel.<>`?

Reply:
It's a method defined on data so that wouldn't work, but you could define a new operator via an implicit conversion:

```scala
object CompatConnect {
  import Chisel._
  implicit class sketchyBulkConnectable(data: Data) {
    def <*>(that: Data): Unit = data <> that
  }
}
```
You can then `import CompatConnect._` in the code that needs it.

Reply:
Fundamentally, relying on this behavior was considered dangerous because in refactoring code if you added a field it would bulk connect anyway and lead to subtle bugs. I think there is a very strong argument that such behavior should work when used with inheritance, eg. if `Foo` and `Bar` are `Bundles` that both extends a common interface `Fizz`, `<>` should be legal between them and only connect the common fields in `Fizz`. Unfortunately, it's really not clear how to implement such a feature (well, probably doable with incredibly gnarly reflection), so it's been languishing.

Reply:
I agree that this is the desired/intended behavior in most extant use cases